### PR TITLE
Improve resiliency external metrics server Part 1

### PR DESCRIPTION
**Pull request by CharlyF**
### What does this PR do?

This is the first main change to go towards improving the resilience of the connection of controllers to the External Metrics Server. 
Starting by bumping the dependency to external repos which were blocking the upgrade of the dependency to Kubernetes.
Then moving to a dynamic client for the WatermarkPodAutoscaler controller.
Next steps are:
- Bump the WPA repo to remove the dependency on k8s/k8s
- Expose configuration parameters of the dynamic client to make connections more resilient 

### Additional Notes

This bumps the following deps:
- datadog-operator to 0.3.1
- wpa to 0.3.0 (not released yet, it's 0.2.1-dirty).
- kubernetes to 1.18.2
- ksm to v1.9.8 (unreleased, but is as close to 2.0.0 as you can get. Includes our custom changes to the builder).
- remove dependency on goautoneg

I had to update a few tests, especially in the LE (not sure how the test passed before). 
Here are a few issues that helped me with the migration to a dynamic client:
- https://github.com/kubernetes/client-go/issues/913
- https://github.com/kubernetes-sigs/kubebuilder/issues/491#issuecomment-459474907
- https://erwinvaneyk.nl/kubernetes-unstructured-to-typed/
- https://sourcegraph.com/github.com/kubernetes/client-go/-/commit/abf396d787a7442759dd7d9315005bedc639e134?visible=2#diff-18c52f60833cb83c20a57cb83518baaf
- https://hackernoon.com/platforms-on-k8s-with-golang-watch-any-crd-0v2o3z1q

### Describe your test plan

Create/Update/Delete WPAs as well as DatadogMetrics and make sure the Cluster Agent processes them correctly.
Check against several version of Kubernetes.

### CODE REVIEW

File: pkg/clusteragent/admission/controllers/secret/controller.go [hunk 1]
```diff
@@ -203,7 +204,7 @@ func (c *Controller) updateSecret(secret *corev1.Secret) error {
 
 	secret = secret.DeepCopy()
 	secret.Data = data
-	_, err = c.clientSet.CoreV1().Secrets(c.config.GetNs()).Update(secret)
+	_, err = c.clientSet.CoreV1().Secrets(c.config.GetNs()).Update(context.TODO(), secret, metav1.UpdateOptions{})
```
Comment (xornivore) [inline pkg/clusteragent/admission/controllers/secret/controller.go:207]: Should we introduce a timeout on this context here? Basically use `context.WithTimeout(..)` here and in similar spots.

File: pkg/clusteragent/externalmetrics/autoscaler_watcher.go [hunk 2]
```diff
@@ -238,12 +245,18 @@ func (w *AutoscalerWatcher) getAutoscalerReferences() (map[string]*externalMetri
...
 		if err != nil {
 			return nil, fmt.Errorf("Could not list WPAs (to update DatadogMetric active status): %v", err)
 		}
 
-		for _, wpa := range wpaList {
+		for _, wpaObj := range wpaList {
+			wpa := &v1alpha1.WatermarkPodAutoscaler{}
+			err := apiserver.StructureIntoWPA(wpaObj, wpa)
+			if err != nil {
+				log.Errorf("Error converting wpa from the cache")
```
Comment (xornivore) [inline pkg/clusteragent/externalmetrics/autoscaler_watcher.go:257]: Mind adding details of the error to this log trace here?

File: pkg/util/kubernetes/apiserver/controller_util.go [hunk 3]
```diff
@@ -122,7 +131,7 @@ func (h *AutoscalersController) gc() {
 		return
 	}
 	h.deleteFromLocalStore(toDelete.External)
-	log.Debugf("Done GC run. Deleted %d metrics", len(toDelete.External))
+	log.Infof("Done GC run. Deleted %d metrics", len(toDelete.External))
```
Comment (xornivore) [inline pkg/util/kubernetes/apiserver/controller_util.go:134]: Do we expose this as expvar?

File: pkg/util/kubernetes/apiserver/util.go [hunk 4]
```diff
@@ -40,3 +44,11 @@ func SyncInformers(informers map[InformerName]cache.SharedInformer) error {
 	}
 	return g.Wait()
 }
+
+func StructureIntoWPA(obj interface{}, structDest *v1alpha1.WatermarkPodAutoscaler) error {
+	unstrObj, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("Could not cast Unstructured object: %v", obj)
```
Comment (xornivore) [inline pkg/util/kubernetes/apiserver/util.go:51]:
```suggestion
		return fmt.Errorf("could not cast Unstructured object: %v", obj)
```

**Review by xornivore — COMMENTED** [review]
Looks great, some small comments.

File: pkg/util/kubernetes/apiserver/wpa_controller.go [hunk 5]
```diff
@@ -178,12 +189,22 @@ func (h *AutoscalersController) syncWPA(key interface{}) error {
...
-	wpaCached, err := h.wpaLister.WatermarkPodAutoscalers(ns).Get(name)
+	wpaCachedObj, err := h.wpaLister.ByNamespace(ns).Get(name)
+	if err != nil {
+		log.Errorf("Could not retrieve key %s from cache: %v", key, err)
+		return err
+	}
+	wpaCached := &apis_v1alpha1.WatermarkPodAutoscaler{}
+	err = StructureIntoWPA(wpaCachedObj, wpaCached)
+	if err != nil {
+		log.Errorf("Could cast wpa %s retrieved from cache to wpa structure: %v", key, err)
```
Comment (xornivore) [inline pkg/util/kubernetes/apiserver/wpa_controller.go:200]:
```suggestion
		log.Errorf("Could not cast wpa %s retrieved from cache to wpa structure: %v", key, err)
```

**Review by xornivore — APPROVED** [review]
👍

**Comment by vboulineau** [comment]
We need to check performances (CPU/Memory) impact as moving to Informers based on DynamicClient may have an impact: https://github.com/kubernetes-sigs/kubebuilder/issues/1152#issuecomment-643440407

**Pull request merged**